### ISSUE TITLE
`no-duplicate-type-constituents` reports explicit `undefined` on optional parameter properties, which typescript-eslint does not

### ISSUE THREAD

**Issue by wagenet** [oxc-project/tsgolint#1221]
_Identified and written by Claude (Claude Code)._

The "Explicit undefined is unnecessary on an optional parameter" check fires on constructor parameter properties (`readonly x?: T | undefined`). `@typescript-eslint/no-duplicate-type-constituents` reports the same shape on an ordinary optional parameter but not on a parameter property.

The difference matters because a parameter property is also a class property, and emitting its declaration depends on that annotation in some toolchains.

## Reproduction

`a.ts`:

```ts
export class A {
	constructor(readonly name?: string | undefined) {}
}
export function f(name?: string | undefined): string | undefined {
	return name;
}
```

`.oxlintrc.json`:

```json
{ "plugins": ["typescript"], "categories": { "correctness": "off" }, "rules": { "typescript/no-duplicate-type-constituents": "error" } }
```

```console
$ oxlint --type-aware a.ts
a.ts:2:39: error typescript(no-duplicate-type-constituents): Explicit undefined is unnecessary on an optional parameter.
a.ts:4:35: error typescript(no-duplicate-type-constituents): Explicit undefined is unnecessary on an optional parameter.
```

ESLint with `@typescript-eslint/no-duplicate-type-constituents: 'error'` on the same file:

```console
  4:35  error  Explicit undefined is unnecessary on an optional parameter  @typescript-eslint/no-duplicate-type-constituents
```

Only the plain parameter. The result is the same with `isolatedDeclarations` on or off.

## Why the parameter-property case should be left alone

Removing that `| undefined` broke our declaration build. Our packages emit `.d.ts` through tsdown with `isolatedDeclarations`, which uses oxc's isolated-declarations emitter, and for `constructor(readonly p?: P)` with `P` a type alias that emitter raised:

```
TS9025: Declaration emit for this parameter requires implicitly adding undefined to it's type. This is not supported with --isolatedDeclarations.
```

That emitter behavior has since been fixed in oxc (oxc-project/oxc#25909, first released in oxc 0.146.0), and `tsc` accepts the file either way, so the build breakage was version-dependent. The rule divergence is not: typescript-eslint leaves parameter properties alone, and anyone following tsgolint's finding while on an older toolchain gets a failing declaration build.

## Impact

Found migrating a monorepo from ESLint to oxlint. Following the two findings this produced, and removing `| undefined` from two parameter properties, took a library's declaration build down, and the rest of the repo then resolved that library to `any`.

## Versions

- oxlint 1.83.0, tsgolint built from `main` (a0e243e)
- `@typescript-eslint/eslint-plugin` 8.67.0, ESLint 10.8.1, TypeScript 6.0.3 for the comparison
- macOS arm64
